### ISSUE TITLE
sparse-training gets contrary  BN distribution 

### ISSUE THREAD

**Issue by summer-wind-hi** [liuzhuang13/slimming#11]
Hi @liuzhuang13,
   
     After sparse training, I visualize histogram distributions of  all BN scaling factors in our yolov3 model which is used to detect  football,   I found a weird phenomenon--I  didn't get sparser BN scaling factors(more factors equals to 0 or near to 0) but far away from 0  compared with base  model, here is the visualization figs:
![image](https://user-images.githubusercontent.com/80259625/119967692-39025b00-bfdf-11eb-929f-7cdca2bfa189.png)

implementation of sparse regularization is continuous with your code ,  I am confused about it and expect your  recommends sincerely , and pls tell me  if there is any misunderstanding , thanks.   

regards.
summer,Gao

**Comment by liuzhuang13** [comment]
Can you try diagnosis with setting the sparsity regularization very large? In that case, all factors should be close to 0, otherwise there might be some issues in the way it's implemented.

**Comment by summer-wind-hi** [comment]
Hi @liuzhuang13,
    Thanks for your prompt reply!
    As your recommend, I increased coefficient of  sparse regularization(trade-off lambda) from 0.0001 to 0.01 and 0.1, the sparse training result was not improved, here is the histogram distribution of BN scale factors :
 
![image](https://user-images.githubusercontent.com/80259625/120058442-70bfe000-c07d-11eb-9468-1674744737b7.png)

they are much farther  away from 0, which  is contrary to our initial  intention.

   As for code, I implemented  and trained both in darknet:
       1) implement subgradient of absolute value function (same as torch.sign())
       2) update BN gradient before  parameters update
   
   here is the code, add following 2 functions before update BN parameters:
       sign_gpu(l.n, l.scales_gpu, 1, l.scale_sign_gpu, 1);
       axpy_gpu(l.n, lambda, l.scale_sign_gpu, 1, l.scale_updates_gpu, 1);
   the cuda implementation of self-defined function sign_gpu() is as following:
        extern "C" void sign_gpu(int N, float * X, int INCX, float * Y, int INCY)
        {
            **sign_kernel**<<<cuda_gridsize(N), BLOCK>>>(N, X, INCX, Y, INCY);
            check_error(cudaPeekAtLastError());
        }
        
      and : 

        __global__ void **sign_kernel**(int N, float *X, int INCX, float *Y, int INCY)
        {
            int i = (blockIdx.x + blockIdx.y*gridDim.x) * blockDim.x + threadIdx.x;
            if(i < N) {
                if(X[i*INCX] > 0) Y[i*INCY] = 1;
                else{
                    if(X[i*INCX] < 0) Y[i*INCY] = -1;
                    else Y[i*INCY] = 0;
                }
            }
        }

      I have  checked my implementation  and didn't find issue,  is there anything I misunderstand? Expect for your advice sincerely , thanks.

regards,
summer, Gao

**Comment by liuzhuang13** [comment]
Is there a reason to code the additional sub-gradient in cuda, instead of in PyTorch/TF? Sorry I'm not that familiar with cuda coding.  It should be easier in pytorch, and you can indeed add a L1 loss into the original loss function instead of manually add the sign function into the gradients.

**Comment by summer-wind-hi** [comment]
Thanks for your quick reply!
1) The reason why code the additional sub-gradient in cuda is that we have been training our yolov3 model in darknet  and the first thing comes to my mind is to implement in darknet, we can compare the result  with our baseline from precision and time comsuption aspects;
2) Implement in pytorch is indeed  easier and I will explore yolov3  pytorch training;

Thanks for your suggestions  and reply！ 

Best regards,
summer, Gao

**Comment by summer-wind-hi** [comment]
Fix the issue.  
It's the implementation issue  because the bn.gradient I get  before update parameters  in darknet is negative,  now sparse training result is as we expect  (sparser) after multiplying -1 to sign(bn.weight).
